### PR TITLE
[TOPIC-BLE-LLCP] Data Length Update procedure

### PR DESCRIPTION
**Pull request by erbr-ot**
First cut of procedure implementation.

Unittest pass, of cause, as well as EDTT/LL tests related to DLU procedure 1M_PHY case (con/sla/bv-77 & 78 and con/mas/bv-73 & 74). The LL tests require a slight rework and aligning this will be in separate PR

### CODE REVIEW

File: subsys/bluetooth/controller/ll_sw/ull_llcp_common.c [hunk 1]
```diff
@@ -323,7 +388,23 @@ static void lp_comm_st_idle(struct ll_conn *conn, struct proc_ctx *ctx, uint8_t
 
 static void lp_comm_st_wait_tx(struct ll_conn *conn, struct proc_ctx *ctx, uint8_t evt, void *param)
 {
-	/* TODO */
+	switch (evt) {
+	case LP_COMMON_EVT_RUN:
+		switch (ctx->proc) {
+		case PROC_DATA_LENGTH_UPDATE:
+			if (!ctx->pause && !tx_alloc_is_available()) {
```
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_common.c:395]: Correct - it's wrong! Should not be negated

File: subsys/bluetooth/controller/ll_sw/ull_llcp_common.c [hunk 2]
```diff
@@ -242,6 +260,32 @@ static void lp_comm_complete(struct ll_conn *conn, struct proc_ctx *ctx, uint8_t
...
+			} else {
+				if (dle_changed) {
+					lp_comm_ntf(conn, ctx);
+				}
+				lr_complete(conn);
+				ctx->state = LP_COMMON_STATE_IDLE;
+			}
+		} else {
+			// Peer does not accept DLU, so disable on current connection
+			feature_unmask_features(conn, LL_FEAT_BIT_DLE);
```
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_common.c:280]: Yeap - needs an lr_complete and setting state idle

File: subsys/bluetooth/controller/ll_sw/ull_llcp_common.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_common.c]: Actually it is cleaner to remove the 'tx_alloc_is_available' part of the clause at let the send_req() func handle this

**Review by kruithofa — APPROVED** [review]

**Review by thoh-ot — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/lll_conn.h [hunk 3]
```diff
@@ -69,16 +78,28 @@ struct lll_conn {
 		uint32_t window_size_event_us;
 	} slave;
 #endif /* CONFIG_BT_PERIPHERAL */
-
```
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/lll_conn.h:72]: check

File: subsys/bluetooth/controller/ll_sw/lll_conn.h [hunk 4]
```diff
@@ -69,16 +78,28 @@ struct lll_conn {
...
 	uint16_t max_tx_time;
 	uint16_t max_rx_time;
 #endif /* CONFIG_BT_CTLR_PHY */
+
+#else /* CONFIG_BT_LL_SW_SPLIT_LLCP_LEGACY */
+	struct {
+		struct dle_t local;
+		struct dle_t remote;
+		struct dle_t eff;
+		uint8_t request_dlu;
```
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/lll_conn.h]: ok

File: subsys/bluetooth/controller/ll_sw/lll_conn.h [hunk 5]
```diff
@@ -69,16 +78,28 @@ struct lll_conn {
...
+		uint8_t ntf_req;
+		uint8_t ntf_ack;
```
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/lll_conn.h]: garbage - removed

File: subsys/bluetooth/controller/ll_sw/openisa/lll/lll_conn.c [hunk 6]
```diff
@@ -422,7 +422,7 @@ void lll_conn_rx_pkt_set(struct lll_conn *lll)
 	LL_ASSERT(node_rx);
 
 #if defined(CONFIG_BT_CTLR_DATA_LENGTH)
-	max_rx_octets = lll->max_rx_octets;
+	max_rx_octets = lll->dle.eff.max_rx_octets;
```
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/openisa/lll/lll_conn.c]: missed it - good eye!

File: subsys/bluetooth/controller/ll_sw/openisa/lll/lll_conn.c [hunk 7]
```diff
@@ -456,7 +456,7 @@ void lll_conn_tx_pkt_set(struct lll_conn *lll, struct pdu_data *pdu_data_tx)
 	uint8_t phy, flags;
 
 #if defined(CONFIG_BT_CTLR_DATA_LENGTH)
-	max_tx_octets = lll->max_tx_octets;
+	max_tx_octets = lll->dle.eff.max_tx_octets;
```
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/openisa/lll/lll_conn.c]: ditto

File: subsys/bluetooth/controller/ll_sw/lll_conn.h [hunk 8]
```diff
@@ -28,6 +28,15 @@ struct node_tx {
 	uint8_t pdu[];
 };
 
+#if defined(CONFIG_BT_CTLR_DATA_LENGTH)
+struct dle_t {
```
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/lll_conn.h]: good idea

**Review by thoh-ot — APPROVED** [review]

**Comment by erbr-ot** [comment]
rebased

**Pull request merged**